src/library/scala/concurrent/Future.scala

```diff
@@ -413,8 +413,34 @@ trait Future[+T] extends Awaitable[T] {
    * @return        a `Future` with the result of the application of `f` to the results of `this` and `that`
    * @group Transformations
    */
-  def zipWith[U, R](that: Future[U])(f: (T, U) => R)(implicit executor: ExecutionContext): Future[R] =
-    flatMap(r1 => that.map(r2 => f(r1, r2)))(if (executor.isInstanceOf[BatchingExecutor]) executor else parasitic)
+  def zipWith[U, R](that: Future[U])(f: (T, U) => R)(implicit executor: ExecutionContext): Future[R] = {
+    val buffer = new AtomicReference[Success[Any]]()
+    val outPromise = Promise[R]()
```

Comment:
@lihaoyi The main drawback of this solution is that now we tye a specific Promise-implementation into the Future-trait, which has been avoided thus far, as such other implementations of Future may now silently switch to a different implementation mid-transformation. Worth discussing

Reply:
I didn't even know you could have other implementations of `Future`; I've never seen or used one personally. Presumably if we had such other implementations, they could override `zipWith` to get whatever behavior they wanted?

Looking at the other "fundamental" methods of future, `def transform` and `def transformWith`, they are basically defined to be "linear" dataflows, so they almost by-definition cannot handle this "whoever completes first" logic that `.zipWith` now provides. In many ways, this new logic is similar to the logic in `Future.firstCompletedOf`, which also uses promises.

If you'd like, we could turn `def zipWith` into a third "fundamental" method, leaving it abstract in `trait Future` and moving the implementation into `DefaultPromise. That may better fit it into the current framework of how `Future` methods are laid out, with a small set of abstract core methods and all other methods just building on top of these.

Reply:
@lihaoyi That's a great point. A good "middle-ground" which would be binary-backwards compatible, would be to leave the implementation in `Future` as is, and then overriding the default implementation in DefaultPromise with the new-updated-behavior. Having the implementation in `DefaultPromise` would also allow us to perform additional optimizations since we know strictly more about `this`.

Reply:
@lihaoyi If we leave the implementation in Future as-is, we can do something like this in DefaultPromise:

```scala
    override final def zipWith[U, R](that: Future[U])(f: (T, U) => R)(implicit executor: ExecutionContext): Future[R] = {
      val state = get()
      if (state.isInstanceOf[Failure[T]]) {
        this.asInstanceOf[Future[R]]
      } else if (state.isInstanceOf[Success[T]]) {
        val l = state.asInstanceOf[Success[T]].get
        that.map(r => f(l, r))
      } else {
        val buffer = new AtomicReference[Success[Any]]()
        val zipped = new DefaultPromise[R]()
        this.onComplete {
          case left: Success[T] =>
            val right = buffer.getAndSet(left).asInstanceOf[Success[U]]
            if (right ne null)
              zipped.tryComplete(try Success(f(left.get, right.get)) catch { case e if NonFatal(e) => Failure(e) })
          case f: Failure[T] =>
            zipped.tryComplete(f.asInstanceOf[Failure[R]])
        }
        that.onComplete {
          case right: Success[U] =>
            val left = buffer.getAndSet(right).asInstanceOf[Success[T]]
            if (left ne null)
              zipped.tryComplete(try Success(f(left.get, right.get)) catch { case e if NonFatal(e) => Failure(e) })
          case f: Failure[U] =>
            zipped.tryComplete(f.asInstanceOf[Failure[R]])
        }
        zipped.future
      }
    }
```